core/services/vrf/listener_v2.go

```diff
@@ -321,141 +384,251 @@ func (lsn *listenerV2) processRequestsPerSub(
-			BlockNumber: m,
-		}, vrfRequest.RequestId)
-		if err != nil {
-			rlog.Warnw("Unable to check if already fulfilled, processing anyways",
-				"err", err,
-				"reqBlockNumber", reqBlockNumber,
-				"currBlock", currBlock,
-			)
-		} else if utils.IsEmpty(callback[:]) {
-			// If seedAndBlockNumber is zero then the response has been fulfilled
-			// and we should skip it
-			rlog.Infow("Request already fulfilled", "callback", callback)
-			lsn.markLogAsConsumed(req.lb)
-			processed[vrfRequest.RequestId.String()] = struct{}{}
-			continue
+		chunk := unconsumed[chunkStart:chunkEnd]
+
+		var unfulfilled []pendingRequest
+		alreadyFulfilled := lsn.checkReqsFulfilled(l, chunk)
+		for i, a := range alreadyFulfilled {
+			if a {
+				lsn.markLogAsConsumed(chunk[i].lb)
+				processed[chunk[i].req.RequestId.String()] = struct{}{}
+			} else {
+				unfulfilled = append(unfulfilled, chunk[i])
+			}
 		}
-		// Run the pipeline to determine the max link that could be billed at maxGasPrice.
-		// The ethcall will error if there is currently insufficient balance onchain.
-		maxLink, run, payload, gaslimit, err := lsn.getMaxLinkForFulfillment(maxGasPriceWei, req, rlog)
+
+		fromAddress, err := lsn.gethks.GetRoundRobinAddress(lsn.chainID, lsn.fromAddresses()...)
 		if err != nil {
-			rlog.Warnw("Unable to get max link for fulfillment, skipping request", "err", err)
+			l.Errorw("Couldn't get next from address", "err", err)
 			continue
 		}
-		if startBalanceNoReserveLink.Cmp(maxLink) < 0 {
-			// Insufficient funds, have to wait for a user top up
-			// leave it unprocessed for now
-			rlog.Infow("Insufficient link balance to fulfill a request, breaking", "maxLink", maxLink)
-			break
-		}
-		rlog.Infow("Enqueuing fulfillment")
-		// We have enough balance to service it, lets enqueue for txm
-		err = lsn.q.Transaction(func(tx pg.Queryer) error {
-			if err = lsn.pipelineRunner.InsertFinishedRun(&run, true, pg.WithQueryer(tx)); err != nil {
-				return err
+		maxGasPriceWei := lsn.cfg.KeySpecificMaxGasPriceWei(fromAddress)
+
+		pipelines := lsn.runPipelines(l, maxGasPriceWei, unfulfilled)
+		for _, p := range pipelines {
+			ll := l.With("reqID", p.req.req.RequestId.String(),
+				"txHash", p.req.req.Raw.TxHash,
+				"maxGasPrice", maxGasPriceWei.String(),
+				"fromAddress", fromAddress,
+				"juelsNeeded", p.juelsNeeded.String(),
+				"maxLink", p.maxLink.String(),
+				"gasLimit", p.gasLimit)
+
+			if p.err != nil {
+				ll.Errorw("Pipeline error", "err", p.err)
+				continue
 			}
-			if err = lsn.logBroadcaster.MarkConsumed(req.lb, pg.WithQueryer(tx)); err != nil {
+
+			if startBalanceNoReserveLink.Cmp(p.maxLink) < 0 {
+				// Insufficient funds, have to wait for a user top up. Leave it unprocessed for now
+				ll.Infow("Insufficient link balance to fulfill a request, returning")
+				return processed
+			}
+
+			ll.Infow("Enqueuing fulfillment")
+			err = lsn.q.Transaction(func(tx pg.Queryer) error {
```

Comment:
Idea, but not sure if it's good: what if we loop through `pipelines` and filter out the ones that failed, and of the remaining ones, we do a `CreateEthTransactionBatch` (would have to be a bptxm side function as well) and we can also do batch updates and inserts of the log consumed and the finished runs (similar to what's in the batch vrf coordinator PR). 

Idea here being we could probably save a lot of time on DB queries if we do all of these in one big transaction, rather than many small ones. 

However, we don't really have database call timings, at least AFAIK, so we'd need to profile and see how much of a speedup we get there.

Reply:
Yeah, could be worth it if creating transactions ends up being a bottleneck, but I kinda don't think it is right now. I think CreateEthTransactionBatch would be kinda hard to write since it's doing a few SQL queries under the hood